app/controllers/users/registrations_controller.rb

```diff
@@ -1,10 +1,41 @@
-# frozen_string_literal: true
-
 class Users::RegistrationsController < Devise::RegistrationsController
+  alias :rebuild_devise_user :build_resource
+
+  def create
+    if email_matches_whitelist
+      do_default_devise_behaviour
+    else
+      rebuild_devise_user
+      add_error_to_devise_user
+      render 'new'
+    end
+  end
+
 protected
 
   # The path used after sign up for inactive accounts.
   def after_inactive_sign_up_path_for(_resource)
     users_confirmations_pending_path
   end
+
+  def do_default_devise_behaviour
+    Devise::RegistrationsController
+      .instance_method(:create).bind(self).call
```

Comment:
This feels a bit like we're forcing our way through this flow.
Happy to discuss

Reply:
Yeah, this might be another indicator that I should validate on the model instead (as [@tommotaylor suggested](https://github.com/alphagov/govwifi-admin/pull/23#discussion_r209229910)) - I'm gonna give that a try.